src/js/plugin.js

```diff
@@ -0,0 +1,436 @@
+   *        event.
+   */
+  handleStateChanged(e) {}
+
+  /**
+   * Disposes a plugin.
+   *
+   * Subclasses can override this if they want, but for the sake of safety,
+   * it's probably best to subscribe the "dispose" event.
+   */
+  dispose() {
+    const {name, player} = this;
+
+    /**
+     * Signals that a class-based plugin is about to be disposed.
+     *
+     * @event dispose
+     * @memberof Plugin
+     * @instance
+     */
+    this.trigger('dispose');
+    this.off();
+
+    // Eliminate any possible sources of leaking memory by clearing up
+    // references between the player and the plugin instance and nulling out
+    // the plugin's state and replacing methods with a function that throws.
+    player[PLUGIN_CACHE_KEY][name] = false;
+    this.player = this.state = null;
+
+    this.dispose = () => {
+      throw new Error('cannot call methods on a disposed object');
+    };
+
+    this.setState = this.off = this.on = this.one = this.trigger = this.dispose;
+
+    // Finally, replace the plugin name on the player with a new factory
+    // function, so that the plugin is ready to be set up again.
+    player[name] = createPluginFactory(name, pluginStorage[name]);
+  }
+
+  /**
+   * Determines if a plugin is a basic plugin (i.e. not a sub-class of `Plugin`).
+   *
+   * @param  {string|Function} plugin
+   *         If a string, matches the name of a plugin. If a function, will be
+   *         tested directly.
+   *
+   * @return {boolean}
+   *         Whether or not a plugin is a basic plugin.
+   */
+  static isBasic(plugin) {
+    const p = (typeof plugin === 'string') ? getPlugin(plugin) : plugin;
+
+    return typeof p === 'function' && !Plugin.prototype.isPrototypeOf(p.prototype);
+  }
+
+  /**
+   * Register a Video.js plugin.
+   *
+   * @param  {string} name
+   *         The name of the plugin to be registered. Must be a string and
+   *         must not match an existing plugin or a method on the `Player`
+   *         prototype.
+   *
+   * @param  {Function} plugin
+   *         A sub-class of `Plugin` or a function for basic plugins.
+   *
+   * @return {Function}
+   *         For class-based plugins, a factory function for that plugin. For
+   *         basic plugins, a wrapper function that initializes the plugin.
+   */
+  static registerPlugin(name, plugin) {
+    if (typeof name !== 'string') {
+      throw new Error(`illegal plugin name, "${name}", must be a string, was ${typeof name}`);
+    }
+
+    if (pluginExists(name) || Player.prototype.hasOwnProperty(name)) {
+      throw new Error(`illegal plugin name, "${name}", already exists`);
+    }
+
+    if (typeof plugin !== 'function') {
+      throw new Error(`illegal plugin for "${name}", must be a function, was ${typeof plugin}`);
+    }
+
+    pluginStorage[name] = plugin;
+
+    // Add a player prototype method for all sub-classed plugins (but not for
+    // the base Plugin class).
```

Comment:
I have two ideas here, maybe the API change would be too drastic, but here they are anyway:

would it be better to just make plugin registration work more like Components, ie: store Plugins that are currently in use on Plugin.plugins_[name]. Then force the user to initialize plugins through something like player.turnPluginOn('name');.
Just initialize right when the player starts (before ready). Then the plugin can attach to player events to do things that it needs to. Then the only way to get a plugin (which will always be an instance?) would be through player.getPlugin